File: LicensingService/src/main/java/SpringMicroservicesInAction/LicensingService/utils/UserContextHolder.java
package SpringMicroservicesInAction.LicensingService.utils;

import org.springframework.util.Assert;

public class UserContextHolder {

    private static final ThreadLocal<UserContext> userContext = new ThreadLocal<UserContext>();

    public static final UserContext getContext() {
        UserContext context = userContext.get();

        if (context == null) {
            context = createEmptyContext();
            userContext.set(context);
        }
        return userContext.get();
    }

    private static final UserContext createEmptyContext() {
        return new UserContext();
    }

    public static final void setContext(UserContext context) {
        Assert.notNull(context, "Only non-null UserContext instances are permitted");
        userContext.set(context);
    }
}

File: OrganizationService/src/main/kotlin/SpringMicroservicesInAction/OrganizationService/utils/UserContextHolder.kt
package SpringMicroservicesInAction.OrganizationService.utils

object UserContextHolder {

    val userContext: ThreadLocal<UserContext> = ThreadLocal()

    fun getContext(): UserContext {
        if (userContext.get() == null) {
            val context = createEmptyContext()
            setContext(context)
        }
        return userContext.get()
    }

    private fun createEmptyContext() = UserContext

    private fun setContext(context: UserContext) {
        userContext.set(context)
    }
}